extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,423 @@
+          .then(this.postShowEndpoint_.bind(this))
+          .then(this.onPostShowEndpointSuccess_.bind(this))
+          .catch(this.onRejection_.bind(this));
+    }
+    return this.getIfHrefPromise_;
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getIfHref_();
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   */
+  show() {
+    this.element.style.display = 'block';
+    this.element.classList.add('amp-active');
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.style.display = 'none';
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_(false);
+    this.postDismissEnpoint_();
+  }
+
+  /** @override */
+  isDismissed() {
+    return this.dialogPromise_;
+  }
+}
+
+
+/**
+ * UserNotificationService handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationService {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDefer>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO: Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    // See https://github.com/ampproject/amphtml/issues/1226
+    onDocumentReady(this.win_, () => {
+      this.documentReadyResolve_();
+    });
+
+    /** @private {!Promise} */
+    this.nextInQueue_ = this.viewer_.isEmbedded() ?
+        this.viewer_.whenVisible() : this.documentReadyPromise_;
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @param {string} id
+   * @return {!Promise}
+   */
+  get(id) {
+    assert(!!this.win_.document.getElementById(id),
+        `No Element found with id ${id} ` +
+        'for `userNotification` service request.');
+    return this.getElementDeferById_(id).promise;
+  }
+
+  /**
+   * Register an instance of `amp-user-notification`.
+   * @param {!UserNotification} userNotification
+   * @package
+   */
+  registerUserNotification(userNotification) {
+    const id = userNotification.getId();
+    const deferred = this.getElementDeferById_(id);
+
+    // Compose the registered notifications into a promise queue
+    // that blocks until one notification is dismissed.
+    this.nextInQueue_ = this.nextInQueue_.then(() => {
+      return userNotification.shouldShow().then((shouldShow) => {
```

Comment:
This is fine, but you might want to move on onRejection here. This is because the functioning of your queue critically depends on the promises never failing. Right now there are many places where each of this promises can fail, thus breaking the queue.